### PR TITLE
[TensorExpr] Block Codegen

### PR DESCRIPTION
**Pull request by protonu**

### CODE REVIEW

File: test/cpp/tensorexpr/test_block.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  LoopNest l({c});
+  For* n_outer;
+  For* n_inner;
+  std::vector<For*> loops = l.getLoopStmtsFor(c);
+  l.splitWithMask(loops[0], block_size, &n_outer, &n_inner);
+  l.setGPUBlockIndex(n_outer, 0);
+  l.setGPUThreadIndex(n_inner, 0);
+  //l.prepareForCodegen();
+  Stmt* stmt = l.root_stmt();
+  BlockCodeGen cuda_cg(stmt, c, a_buf, b_buf);
```
Comment (zheng-xq) [inline test/cpp/tensorexpr/test_block.cpp:38]: Change the name "cuda_cg"

File: test/cpp/tensorexpr/test_block.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+#include "torch/csrc/jit/tensorexpr/loopnest.h"
+#include "torch/csrc/jit/tensorexpr/tensor.h"
+
+#include <c10/util/Half.h>
+
+namespace torch {
+namespace jit {
+using namespace torch::jit::tensorexpr;
+
+static void testBlockTestVectorAdd_impl(int N, int block_size) {
```
Comment (zheng-xq) [inline test/cpp/tensorexpr/test_block.cpp:19]: In addition to check BlockCodeGen can be constructed, it is better to check the generated code matches expectation to some extent.

File: test/test_tensorexpr.py [hunk 3]
```diff
@@ -1260,5 +1260,18 @@ def test(x):
         assert torch.allclose(scripted(a), 2 * a)
         assert cx.elapsed_value() == 1
 
+    def test_simple_add(self):
+        torch._C._jit_set_te_generate_block_code(True)
```
Comment (zheng-xq) [inline test/test_tensorexpr.py:1264]: The standard practice is to remember the old value before, and only revert to that when you are done.

File: test/test_tensorexpr.py [hunk 4]
```diff
@@ -1260,5 +1260,18 @@ def test(x):
...
+
+        def simple(a, b):
+            return torch.add(a, b)
+
+        a = torch.ones(256, 256)
+        b = torch.ones(256, 256)
+        traced = torch.jit.trace(simple, (torch.ones(256, 256), torch.ones(256, 256)))
+        f = traced(a, b)
```
Comment (zheng-xq) [inline test/test_tensorexpr.py:1272]: Check the results?

File: torch/csrc/jit/tensorexpr/block_codegen.cpp [hunk 5]
```diff
@@ -0,0 +1,319 @@
...
+  os() << "arguments {" << std::endl;
+
+  for (auto const& arg : this->dim_values_map) {
+    emitIndent();
+    emitIndent();
+    os() << "var " << arg.first << " = " << arg.second << std::endl;
+  }
+
+  emitIndent();
+  emitIndent();
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/block_codegen.cpp:155]: Why do we need a double indent? If it's needed for correctness, it's totally fine, if it's just for aesthetics, then I suggest using it once (and if you feel strongly about number of spaces, change the `emitIndent` function :) ).

File: torch/csrc/jit/tensorexpr/analysis.h [hunk 6]
```diff
@@ -46,6 +47,102 @@ class NodeFinder : public IRVisitor {
...
+    return block_size_;
+  }
+
+  void analysis_ran() {
+    analysis_runs_++;
+    return;
+  }
+
+  bool areBufsInMap(const std::unordered_set<const Buf*>& bufs) const {
+    for (auto const& arg : bufs) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/analysis.h]: I suggest moving the implementation to a cpp file to keep header files lean.

File: torch/csrc/jit/tensorexpr/block_codegen.cpp [hunk 7]
```diff
@@ -0,0 +1,319 @@
...
+    os() << block_analysis_->getInputName(arg);
+  }
+  for (auto const& arg : buf_reads) {
+    os() << ";" << block_analysis_->getInputName(arg);
+  }
+  os() << ")";
+
+  stmt_v->accept(printer_.get());
+
+#if DEBUG_PRINT
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/block_codegen.cpp:302]: Please use `GRAPH_DEBUG`/`GRAPH_DUMP`/`GRAPH_UPDATE` instead of adhoc debug prints. I know DEBUG_PRINT is used in cuda codegen, but it's because we've been lazy to switch it to the proper logging, not because it should be that way :)

**Review by ZolotukhinM — COMMENTED** [review]
I left a bunch of comments, they are mostly cosmetic. I'll try to look deeper soon.

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 8]
```diff
@@ -267,6 +267,14 @@ c10::optional<Node*> tryMerge(
       getHeader(consumer),
       ":\n");
 
+  // Shortcut for Block code generation for single op
+  if (tensorexpr::getTEGenerateBlockCode() &&
```
Comment (zheng-xq) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:271]: I am fine with this in the short term, but we should be committed to cleaning it up soon. At least add a TODO to yourself to remind you of that.

What I am objecting to is that this code is looking at a global flag to decide whether to generate Block code. This is generally pretty bad. We should only look at input_ device to pick whether this is the target device. SImilar to how CPU and GPU specific codegen is invoked. 

https://github.com/pytorch/pytorch/blob/6bdfd6ae1adb341f5bae2fc8bc987653b8a319ad/torch/csrc/jit/tensorexpr/kernel.cpp#L1392

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 9]
```diff
@@ -267,6 +267,14 @@ c10::optional<Node*> tryMerge(
...
+      producer->kind() == prim::Constant &&
+      (consumer->kind() == aten::add || consumer->kind() == aten::mul)) {
```
Comment (zheng-xq) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:273]: Minor nitpick: I expect this list to get fairly long. So it is better to separate into its own function, and follow similar structure here to make the list easier to view. 

https://github.com/pytorch/pytorch/blob/6bdfd6ae1adb341f5bae2fc8bc987653b8a319ad/torch/csrc/jit/passes/tensorexpr_fuser.cpp#L20

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 10]
```diff
@@ -48,6 +50,10 @@ int& getTECudaPointwiseBlockSize() {
   return te_cuda_pointwise_block_size;
 }
 
+bool& getTEGenerateBlockCode() {
```
Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/kernel.cpp:52]: Look at my other commend on this flag. I would strongly prefer to use the input device, instead of a flag to get into this mode. A TODO might be fine for now to get this unblocked.

File: torch/csrc/jit/tensorexpr/analysis.h [hunk 11]
```diff
@@ -46,6 +47,102 @@ class NodeFinder : public IRVisitor {
...
 };
 
+// A class that analyzes the given program relevant for Block backend.
+class BlockAnalysis : public IRVisitor {
+ public:
+  bool is_buf_store_target(const Buf* buf) const {
+    return store_targets_.count(buf) > 0;
+  }
+
+  std::unordered_set<const Buf*> reads() const {
```
Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/analysis.h]: Do you intend to return by value here?

File: torch/csrc/jit/tensorexpr/analysis.h [hunk 12]
```diff
@@ -46,6 +47,102 @@ class NodeFinder : public IRVisitor {
...
+    return loads_;
+  }
+
+  std::unordered_set<const Buf*> stores() const {
```
Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/analysis.h]: Same as above

File: torch/csrc/jit/tensorexpr/analysis.h [hunk 13]
```diff
@@ -46,6 +47,102 @@ class NodeFinder : public IRVisitor {
...
+  }
+
+  std::string getFlatInputName(const Buf* buf) const {
+    return getInputName(buf) + "_flat";
+  }
+
+
+ private:
+  void visit(const Store* v) override {
+    if (analysis_runs_) {
```
Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/analysis.h]: It seems that you are only using analysis_runs_ as a boolean flag. In that case, it is better to use a "bool analysis_has_run_" flag, instead of a counter. A counter could overflow. Although the likelihood is low, it is better to not have it by design.

File: torch/csrc/jit/tensorexpr/block_codegen.cpp [hunk 14]
```diff
@@ -0,0 +1,319 @@
+#include <torch/csrc/jit/tensorexpr/block_codegen.h>
+
+#include <torch/csrc/jit/tensorexpr/analysis.h>
+#include <torch/csrc/jit/tensorexpr/eval.h>
+#include <torch/csrc/jit/tensorexpr/exceptions.h>
+#include <torch/csrc/jit/tensorexpr/execution_counter.h>
+#include <torch/csrc/jit/tensorexpr/ir_simplifier.h>
+
+#define DEBUG_PRINT 1
```
Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/block_codegen.cpp:9]: Minor: it is generally a bad practice to send out code with DEBUG_PRINT enabled for code review. The standard practice is to enable it locally, and disable it when sending out for code review, so you don't forget to remove it when landing the code. Personally, I often have a local DO_NOT_SUBMIT change that I keep rebasing.

File: torch/csrc/jit/tensorexpr/block_codegen.h [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+namespace torch {
+namespace jit {
+namespace tensorexpr {
+// A class that overrides the underlying IRPrinter to produce Block.
+class BlockPrinter : public IRPrinter {
+ public:
+  explicit BlockPrinter(std::ostream* os, const BlockAnalysis* block_analysis)
+      : IRPrinter(*os), block_analysis_(block_analysis) {}
+  void visit(const For* v) override;
```
Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/block_codegen.h:111]: Minor: virtual functions could be marked as private whenever possible.

**Review by zheng-xq — CHANGES_REQUESTED** [review]

File: torch/csrc/jit/tensorexpr/block_codegen.cpp (hunk 5 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/block_codegen.cpp:153]: Changed

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 10 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/kernel.cpp:55]: I added a TODO for now. I am going to revisit this later. Thanks for the design inputs!

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp (hunk 8 above)
Comment (protonu) [reply torch/csrc/jit/passes/tensorexpr_fuser.cpp:284]: Thanks, this was really useful feedback. I added the TODO for now. I'll investigate adding the Block device type.

File: torch/csrc/jit/tensorexpr/block_codegen.cpp (hunk 7 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/block_codegen.cpp]: done!

File: torch/csrc/jit/tensorexpr/block_codegen.cpp (hunk 5 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/block_codegen.cpp:153]: removed

File: torch/csrc/jit/tensorexpr/analysis.h (hunk 6 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/analysis.h]: Done!

File: torch/csrc/jit/tensorexpr/block_codegen.cpp (hunk 14 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/block_codegen.cpp]: Removed this and using GRAPH_DUMP

File: torch/csrc/jit/tensorexpr/block_codegen.h (hunk 15 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/block_codegen.h]: Thanks for this! Done

File: test/cpp/tensorexpr/test_block.cpp (hunk 1 above)
Comment (protonu) [reply test/cpp/tensorexpr/test_block.cpp]: I removed this outdated unit test for now. I may add a new one later.

File: test/cpp/tensorexpr/test_block.cpp (hunk 2 above)
Comment (protonu) [reply test/cpp/tensorexpr/test_block.cpp]: Same as above; I removed this outdated unit test for now. I may add a new one later.

File: torch/csrc/jit/tensorexpr/analysis.h (hunk 13 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/analysis.h]: Thanks! I did not need this variable anymore (to keep track of anlysis passes).  I removed it.

File: torch/csrc/jit/tensorexpr/analysis.h (hunk 12 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/analysis.h]: Thanks! This was a mistake. Returning by const ref now.

File: torch/csrc/jit/tensorexpr/analysis.h (hunk 11 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/analysis.h]: Thanks! This was a mistake. Returning by const ref now.

File: torch/csrc/jit/tensorexpr/block_codegen.cpp [hunk 16]
```diff
@@ -0,0 +1,319 @@
...
+    case ScalarType::Float:
+      return "2"; // Return Half for now
+    default:
+      return dtype.ToCppString();
+  }
+}
+
+void BlockPrinter::visit(const Add* v) {
+  emitIndent();
+  os() << "add(";
```
[the remark this answers is not in the document]
Comment (protonu) [reply torch/csrc/jit/tensorexpr/block_codegen.cpp:90]: This is the only comment I didn't quite address. I think right now, it should be left as is. In Block, c = a+b, become add(a, b, c). That's the reason, I close the paranthesis on the store. When we support fused ops, I'll revisit this.

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 17]
```diff
@@ -267,6 +278,16 @@ c10::optional<Node*> tryMerge(
       getHeader(consumer),
       ":\n");
 
+  // This is a shortcut for Block code generation for single op.
+  // TODO: This needs to be cleaned up. Block code gen should use
+  // device type and not rely on a global flag.
+  if (tensorexpr::getTEGenerateBlockCode() &&
+      producer->kind() == prim::Constant &&
+      tensorexpr::isSupportedForBlock(consumer)) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:286]: This check probably belongs to `canHandle` function (see line 199 where we check if an op is supported by NNC). Are we doing this only because we cannot really merge ops yet and have to enable "fusion" for a single op? In a longer term it feels like the way fuser constructs subgraphs for fusion should be modified to support single-ops more easily (but if it's too much hassle for now I think it's fine with committing this PR first).

File: torch/csrc/jit/tensorexpr/block_codegen.h [hunk 18]
```diff
@@ -0,0 +1,147 @@
...
+namespace tensorexpr {
+
+// A class that analyzes the given program relevant for Block backend.
+class BlockAnalysis : public IRVisitor {
+ public:
+  bool is_buf_store_target(const Buf* buf) const {
+    return store_targets_.count(buf) > 0;
+  }
+
+  const std::unordered_set<const Buf*>& reads() const {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/block_codegen.h:26]: Nit: should it be "loads" for consistency? :)

File: torch/csrc/jit/tensorexpr/codegen.h [hunk 19]
```diff
@@ -49,6 +49,10 @@ class TORCH_API CodeGen {
     return device_;
   }
 
+  virtual std::string getCodeText() {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/codegen.h:52]: What does it do?

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 20]
```diff
@@ -1284,11 +1321,14 @@ TensorExprKernel::BackendType TensorExprKernel::inferBackendTypeFromDevice(
...
   if (device.type() == at::kCUDA) {
     backendType = kCudaCodeGen;
+  } else if (device.type() == at::kCPU && getTEGenerateBlockCode()) {
+    backendType = kBlockCodeGen;
   } else if (device.type() == at::kCPU) {
 #ifdef TORCH_ENABLE_LLVM
     backendType = kLLVMCodeGen;
 #else
     backendType = kSimpleIREval;
+    std::cout << "device also used\n";
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:1331]: Leftover?

**Review by ZolotukhinM — APPROVED** [review]
Looks good to me overall!

File: torch/csrc/jit/tensorexpr/block_codegen.h (hunk 18 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/block_codegen.h:26]: Makes a lot of sense :) Will change it

File: torch/csrc/jit/tensorexpr/kernel.cpp (hunk 20 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/kernel.cpp:1331]: Argh! forgot this. Thanks for catching it, will remove.

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp (hunk 17 above)
Comment (protonu) [reply torch/csrc/jit/passes/tensorexpr_fuser.cpp:286]: Yes, that's right. We aren't merging ops for Block right now, and we needed a way to pass in a single op. When we starting supporting fusion, I'll modify this code.

File: torch/csrc/jit/tensorexpr/codegen.h (hunk 19 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/codegen.h:52]: Glow needed to get generated code from NNC (in this case Block codegen), which it would then compile. This was a hook for that.  Glow had an access to codegen but not the exact type of codegen.

Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/codegen.h:52]: I see, makes sense. Could you please add a comment for this function?

Comment (ZolotukhinM) [reply torch/csrc/jit/tensorexpr/codegen.h:52]: In a longer term I think it probably should be generalized (like, it could return `void*`, for instance, and give some representation of the code produced by the codegen - it could be a string, but doesn't have to in a general case). Returning a pointer to the codegen itself also might be a viable option I think.

Comment (protonu) [reply torch/csrc/jit/tensorexpr/codegen.h:52]: That makes a lot of sense! I'll look into this

Comment (protonu) [reply torch/csrc/jit/tensorexpr/codegen.h:52]: Right now for the current pipeline we did return the pointer to the codegen.

Comment (protonu) [reply torch/csrc/jit/tensorexpr/codegen.h:52]: I should also rename this function. Will add that in the TODO.

File: torch/csrc/jit/tensorexpr/analysis.h [hunk 21]
```diff
@@ -46,6 +47,31 @@ class NodeFinder : public IRVisitor {
   std::vector<Node*> nodes;
 };
 
+// A class that analyzes the given program relevant for Block backend
+// It creates a map of multi dim buffers and their flat verions
+class CreateBufferMap : public IRVisitor {
+ public:
+  std::unordered_map<std::string, const Buf*> getBufferMap() const {
```
Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/analysis.h:54]: Same here. Why return by value, instead of const reference?

File: torch/csrc/jit/tensorexpr/block_codegen.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (zheng-xq) [reply torch/csrc/jit/tensorexpr/block_codegen.cpp:90]: I am fine with this change for now. But what should really happen here is to first mutate that expression into the right statement, and print it out in the right format. It is unclear from this comment alone, how you would address "x = a + b + c". 

A comment to clarify why it behaves like this is helpful. Also a TODO to clean it up later will be good as well.

File: torch/csrc/jit/tensorexpr/block_codegen.h [hunk 22]
```diff
@@ -0,0 +1,147 @@
...
+  std::unordered_map<std::string, const Buf*> map_input_to_tensor_bufs_;
+  std::unordered_set<const Buf*> store_targets_;
+  std::unordered_set<const Buf*> loads_;
+  int block_size_ = 32;
+};
+
+// A class that overrides the underlying IRPrinter to produce Block.
+class BlockPrinter : public IRPrinter {
+ public:
+  explicit BlockPrinter(std::ostream* os, const BlockAnalysis* block_analysis)
```
Comment (zheng-xq) [inline torch/csrc/jit/tensorexpr/block_codegen.h:66]: Nit picking: explicit is not useful here.

**Review by zheng-xq — APPROVED** [review]
LGTM. Please address my current set of comments before submitting. Thanks

File: torch/csrc/jit/tensorexpr/analysis.h (hunk 21 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/analysis.h:54]: Thanks! fixed

File: torch/csrc/jit/tensorexpr/block_codegen.h (hunk 22 above)
Comment (protonu) [reply torch/csrc/jit/tensorexpr/block_codegen.h:66]: thanks, took it out

File: torch/csrc/jit/tensorexpr/block_codegen.cpp (hunk 16 above)
[the remark this answers is not in the document]
Comment (protonu) [reply torch/csrc/jit/tensorexpr/block_codegen.cpp:90]: Thanks, that's great feedback.
I'll add the TODO for now. I am going to follow up on mutating the expression when I start handling fused ops.

File: torch/csrc/jit/passes/tensorexpr_fuser.cpp [hunk 23]
```diff
@@ -278,7 +293,7 @@ class TensorExprFuser {
       it++;
     }
 
-    for (Node* n : fusion_groups) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/tensorexpr_fuser.cpp:281]: Nit: unnecessary change (and the one above that).

File: torch/csrc/jit/tensorexpr/analysis.h [hunk 24]
```diff
@@ -67,6 +68,31 @@ class VarFinder : public IRVisitor {
...
+
+ private:
+  void visit(const Store* v) override {
+    auto load_node = dynamic_cast<const Load*>(v->value());
+    auto call_node = dynamic_cast<const FunctionCall*>(v->value());
+    if (load_node || call_node) {
+      TORCH_INTERNAL_ASSERT(!(load_node && call_node));
+      auto t_buf = load_node ? load_node->buf() : call_node->tensor()->buf();
+      if (load_node) {
+        map_input_to_tensor_bufs_.emplace(t_buf->name_hint(), v->buf());
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/analysis.h:87]: I'm not sure it's a good idea to use 'name_hint' as a key. I think we never tried to guarantee that they were unique.

File: torch/csrc/jit/tensorexpr/kernel.cpp [hunk 25]
```diff
@@ -17,7 +17,8 @@ namespace tensorexpr {
 static int te_cuda_pointwise_loop_levels = -1;
 static int te_cuda_pointwise_block_count = -1;
 static int te_cuda_pointwise_block_size = -1;
-static bool fallback_allowed = false;
+static bool fallback_allowed = true;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/tensorexpr/kernel.cpp:20]: Please undo this change :)
